lib/puppet_x/puppetlabs/aws.rb

```diff
@@ -56,21 +56,85 @@ def self.read_only(*methods)
       end
 
       def self.logger
-        if ENV['PUPPET_AWS_DEBUG_LOG'] and not ENV['PUPPET_AWS_DEBUG_LOG'].empty?
-          Logger.new('puppet-aws-debug.log')
+        log_name = 'puppet-aws-debug.log'
+        if global_configuration and global_configuration['default'] and global_configuration['default']['logger']
+          Logger.new(log_name) if global_configuration['default']['logger'] == 'true'
+        elsif ENV['PUPPET_AWS_DEBUG_LOG'] and not ENV['PUPPET_AWS_DEBUG_LOG'].empty?
+          Logger.new(log_name)
+        else
+          nil
+        end
+      end
+
+      def self.global_credentials
+        # Under a Puppet agent we don't have the HOME environment variable available
+        # so the standard way of detecting the location for the config file doesn't
+        # work. The following provides a fall back method to a confdir config file.
+        # The preference is still to use IAM instance roles if possible.
+        begin
+          Puppet.initialize_settings unless Puppet[:confdir]
+          path = File.join(Puppet[:confdir], 'puppetlabs_aws_credentials.ini')
+          credentials = ::Aws::SharedCredentials.new(path: path)
+          credentials.loadable? ? credentials : nil
+        rescue ::Aws::Errors::NoSuchProfileError
+          nil
+        end
+      end
+
+      def self.global_configuration
+        Puppet.initialize_settings unless Puppet[:confdir]
+        path = File.join(Puppet[:confdir], 'puppetlabs_aws_configuration.ini')
+        File.exists?(path) ? parse_ini(File.new(path)) : nil
+      end
+
+      def self.region_from_global_configuration
+        global_configuration['default']['region']
+      end
+
+      def self.proxy_configuration
+        if global_configuration and global_configuration['default'] and global_configuration['default']['http_proxy']
+          global_configuration['default']['http_proxy']
+        elsif ENV['PUPPET_AWS_PROXY'] and not ENV['PUPPET_AWS_PROXY'].empty?
+          ENV['PUPPET_AWS_PROXY']
         else
           nil
         end
       end
 
       def self.client_config(region)
-        config = {region: region, logger: logger}
-        if ENV['PUPPET_AWS_PROXY'] and not ENV['PUPPET_AWS_PROXY'].empty?
-          config[:http_proxy] = ENV['PUPPET_AWS_PROXY']
+        config = {logger: logger}
+        config[:http_proxy] = proxy_configuration if proxy_configuration
+        config[:credentials] = global_credentials if global_credentials
+        config[:region] = if global_configuration
+          config[:region] = region_from_global_configuration || region
+        else
+          region
         end
         config
       end
 
+      # This method is vendored from the AWS SDK, rather than including an
+      # extra library just to parse an ini file
```

Comment:
Just to make sure: this is not part of the aws-sdk-core gem this currently requires, and adding this as a runtime dependency would require all users to install additional code on nodes running the module?